lib/protoflow-blocks/src/blocks/text/encode_csv.rs

```diff
@@ -0,0 +1,135 @@
+};
+use protoflow_derive::Block;
+use simple_mermaid::mermaid;
+use csv::WriterBuilder;
+use std::io::Cursor;
+/// A block that encodes CSV files by converting a header and rows, provided as `prost_types::Value` streams, into a byte stream.
+///
+/// # Block Diagram
+#[doc = mermaid!("../../../doc/text/encode_csv.mmd")]
+///
+/// # Sequence Diagram
+#[doc = mermaid!("../../../doc/text/encode_csv.seq.mmd" framed)]
+///
+/// # Examples
+///
+/// ## Using the block in a system
+///
+/// ```rust
+/// # use protoflow_blocks::*;
+/// # fn main() {
+/// System::build(|s| {
+///     // TODO
+/// });
+/// # }
+/// ```
+///
+/// ## Running the block via the CLI
+///
+/// ```console
+/// $ protoflow execute EncodeCsv
+/// ```
+///
+#[derive(Block, Clone)]
+pub struct EncodeCsv {
+    /// The header message proto_types::Value stream.
+    #[input]
+    pub header: InputPort<Value>,
+    /// The rows message proto_types::Value stream.
+    #[input]
+    pub rows: InputPort<Value>,
+    /// The output message stream.
+    #[output]
+    pub output: OutputPort<Bytes>,
+}
+
+impl EncodeCsv {
+    pub fn new(header: InputPort<Value>, rows: InputPort<Value>, output: OutputPort<Bytes>) -> Self {
+        Self { header, rows, output }
+    }
+
+    pub fn with_system(system: &System) -> Self {
+        use crate::SystemBuilding;
+        Self::new(system.input(), system.input(), system.output())
+    }
+}
+
+impl Block for EncodeCsv {
+    fn execute(&mut self, runtime: &dyn BlockRuntime) -> BlockResult {
+        runtime.wait_for(&self.header)?;
+        runtime.wait_for(&self.rows)?;
+
+        let mut buffer = Cursor::new(Vec::new());
+        {
+            let mut wtr = WriterBuilder::new()
+                .has_headers(false)
+                .from_writer(&mut buffer);
+
+            if let Some(header) = self.header.recv()? {
+                if let Some(ListValue(lv)) = header.kind {
+                    let header_row: Vec<String> = lv.values
+                        .into_iter()
+                        .filter_map(|v| if let Some(StringValue(s)) = v.kind { Some(s) } else { None })
+                        .collect();
+
+                    wtr.write_record(header_row)
+                        .map_err(|e| BlockError::Other(e.to_string()))?;
+                }
+            }
+
+            while let Some(row) = self.rows.recv()? {
+                if let Some(ListValue(lv)) = row.kind {
+                    let row_values: Vec<String> = lv.values
+                        .into_iter()
+                        .filter_map(|v| if let Some(StringValue(s)) = v.kind { Some(s) } else { None })
+                        .collect();
+
+                    wtr.write_record(row_values)
+                        .map_err(|e| BlockError::Other(e.to_string()))?;
+                }
+            }
+
+            wtr.flush().map_err(|e| BlockError::Other(e.to_string()))?;
+        }
+
+        self.output.send(&Bytes::from(buffer.into_inner()))?;
```

Comment:
This is not going to work well. Send rows right when you get them, not buffered up. Otherwise a big CSV file will exhaust memory, and also nothing can flow until the whole file has been parsed and buffered.